src/shared/codelens/pythonCodeLensProvider.ts

```diff
@@ -59,75 +70,210 @@ const getLambdaHandlerCandidates = async ({ uri }: { uri: vscode.Uri }): Promise
+                debugPort: debugPort,
+                outputDir: samProjectCodeRoot,
+            })
+            handlerName = debugHandlerName
+            manifestPath = await makePythonDebugManifest({
+                samProjectCodeRoot,
+                outputDir: baseBuildDir
+            })
+        } else {
+            handlerName = args.handlerName
         }
-        // TODO: Figure out Python specific params and create PythonDebugConfiguration if needed
-        const debugConfig: DebugConfiguration = {
-            type: PYTHON_LANGUAGE,
-            request: 'attach',
-            name: 'SamLocalDebug',
-            preLaunchTask: undefined,
-            address: 'localhost',
-            port: debugPort!,
-            localRoot: samProjectCodeRoot,
-            remoteRoot: '/var/task',
-            skipFiles: []
-        }
+        const inputTemplatePath = await makeInputTemplate({
+            baseBuildDir,
+            codeDir: samProjectCodeRoot,
+            documentUri: args.document.uri,
+            handlerName,
+            runtime,
+            workspaceUri: args.workspaceFolder.uri
+        })
+        logger.info(`pythonCodeLensProvider.initialize: ${
+            JSON.stringify({samProjectCodeRoot, inputTemplatePath, handlerName, manifestPath}, undefined, 2)
+        }`)
+
+        const channelLogger = getChannelLogger(toolkitOutputChannel)
+        const codeDir = samProjectCodeRoot
+        const samTemplatePath: string = await executeSamBuild({
+            baseBuildDir,
+            channelLogger,
+            codeDir,
+            inputTemplatePath,
+            manifestPath,
+            samProcessInvoker: processInvoker,
 
-        const localLambdaRunner: LocalLambdaRunner = new LocalLambdaRunner(
+        })
+    
+        const debugConfig: PythonDebugConfiguration = makeDebugConfig({debugPort, samProjectCodeRoot})
+        await invokeLambdaFunction({
+            baseBuildDir,
+            channelLogger,
             configuration,
-            args,
-            debugPort,
-            runtime,
-            toolkitOutputChannel,
-            processInvoker,
-            taskInvoker,
             debugConfig,
-            samProjectCodeRoot,
-            // TODO: Use onWillAttachDebugger &/or onDidSamBuild to enable debugging support
-        )
-
-        await localLambdaRunner.run()
+            samTaskInvoker: taskInvoker,
+            samTemplatePath,
+            documentUri: args.document.uri,
+            handlerName,
+            isDebug: args.isDebug,
+            onWillAttachDebugger: async () => {
+                // TODO: Find out why debugger can't detach without introducing delay
+                if (process.platform === 'darwin') {
+                    await new Promise(resolve => { // delay to avoid racing condition
+                        logger.info(`pythonCodeLensProvider.initialize ${process.platform} hack: sleeping......`)
+                        setTimeout(resolve, 3000)
```

Comment:
Roughly how much of a delay is needed on your machine? Suggest to take that number, and multiply it by a large safety factor.

Reply:
Perhaps, but that would punish all to make the experience as bad on the best systems as it would be for the lowest performers. If we leave this in (and I hope we don't have to) we may wish to consider:

1. Providing a reasonable default that works for most users 
and
2. Providing an extension configuration option to set this delay for particularly slow ones

Reply:
For me 1.5 seconds works sometimes. 3 works most (I dare not say all) of the time... but ya seems to always work.

Reply:
Can we run some sort of a backoff instead? Something like 1 second -> 2 seconds -> 4 seconds -> 8 seconds -> fail?

Reply:
Added backoff due to being out voted but I feel it's a worse user experience.

Reply:
Retry and back-off disabled for now.